### PR TITLE
Allow declaring iterator in records

### PR DESCRIPTION
**Pull request by YairHalberstadt**
Fixes https://github.com/dotnet/roslyn/issues/46657

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs [hunk 1]
```diff
@@ -19,8 +19,9 @@ internal abstract class SourceConstructorSymbolBase : SourceMemberMethodSymbol
         protected SourceConstructorSymbolBase(
             SourceMemberContainerTypeSymbol containingType,
             Location location,
-            CSharpSyntaxNode syntax)
-            : base(containingType, syntax.GetReference(), ImmutableArray.Create(location), SyntaxFacts.HasYieldOperations(syntax))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs:23]: >SyntaxFacts.HasYieldOperations(syntax) [](start = 91, length = 38)

Thanks much for fixing this. 

nit: I'm wondering if we could add an assertion to `HasYieldOperations` to either say we're not passing in  type declarations, or we're only passing in method/lambda/local function syntaxes, or that it shouldn't find a descendant node that's a method...

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 2]
```diff
@@ -22198,5 +22198,30 @@ private void Handle3(SymbolAnalysisContext context)
...
+            System.Console.Write(i);
+        }
+    }
+    public IEnumerable<int> GetItems() { yield return a; yield return a + 1; }
+}";
+
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "4243");
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:22224]: Add `, verify: Verification.Skipped /* init-only */` to resolve test failure on desktop runtime.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs (hunk 1 above)
Comment (YairHalberstadt) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs:23]: Added Assertion

**Comment by RikkiGibson** [comment]
@jaredpar @jcouv should we backport this to preview2 while the window is open, or just let it go in preview3?

**Pull request merged**